### ISSUE TITLE
Huge memory demand in resquiggle.

### ISSUE THREAD

**Issue by JohnUrban** [nanoporetech/tombo#36]
Hi Marcus,

I'm wondering if you have insight into what exactly is the memory hog in resquiggle (exceeding 67 GB).
```
Wed Feb 21 16:40:21 EST 2018
resquiggle
Getting file list.
Using default canonical ***** DNA ***** model.
Estimating global scale parameter
Correcting 8644 files with 1 subgroup(s)/read(s) each (Will print a dot for each 1000 reads completed).
.slurmstepd: Job 18951741 exceeded memory limit (67181424 > 67108864), being killed
slurmstepd: *** JOB 18951741 ON node850 CANCELLED AT 2018-02-23T20:32:32 ***
slurmstepd: Exceeded job memory limit
```

I am using the much more sensitive minimap index I discussed previously and I did not previously have memory issues with the default index. Therefore, I am thinking it is either:
(i) the mapping step, or 
(ii) a tombo step working with more difficult reads that mapped with the more sensitive index -- if that even makes sense.

Let me know what you think. Many thanks as usual.

Best,

  John

**Comment by marcus1487** [comment]
Hi John,

Yes I recently encountered this issue as well and am working to fix it. I thought I fixed this by converting to Fasta on-disk random access via `pyfaidx` in the last release. But it turns out that the mappy index can actually get quite large for larger genomes and I am currently opening a different `mappy.Aligner` in each process. I have raised this issue in minimap2/mappy (https://github.com/lh3/minimap2/issues/125), but unfortunately the fix is going to involve a bit of a rework in tombo to run mappy in a shared memory space, but running the re-squiggle bits in separate, connected processes to avoid the GIL.

The fix for now is unfortunately to run on fewer processes for large genomes. I hope to have a fix for this soon so that larger genomes can still be run on many cores. I had meant to raise this issue myself, so thank you for raising it!

**Comment by JohnUrban** [comment]
Thanks for getting back. I'm glad it is on your radar - and eagerly await the enhancement.

If the enhancement to tombo will take significant re-working, do you think there is any chance of adding a pick-up-where-you-left-off option? The problem is that I've already reduced the number of processes twice already, and each time it needs to start completely over and has taken a little over 2 days before it crashed.
```Estimating global scale parameter``` seems to be a very long step, but then it goes on to ```Correcting 8644 files with 1 subgroup....``` and crashes soon thereafter. It would be great if the global scale parameter was saved to file (even if temporarily) to be able to pick up after that time-consuming step. I am already guessing there is some RTFM going on here -- so be gentle.  :)

Best,

  John

**Comment by marcus1487** [comment]
No RTFM here really as this is a bit of a buried command, but useful in this case. You can estimate this scale parameter once using the `estimate_scale` command and then provide that result to the `--fixed-scale` option for `resquiggle`. That is exactly what the `Estimating global scale parameter` step is doing. Here is the [bit from the docs](https://nanoporetech.github.io/tombo/model_training.html#estimate-scale), this one is quite buried as I said.

That being said I would venture to guess that this is not actually where most of your time is being spent. From my experience most of the "boot up" time is spent enumerating the list of read files to be processed. The global scale estimation step only opens and processes 1000 files, so in my experience that has not been the bottleneck.

In terms of a "pick up where you left off" feature, that would be a pretty big change as I would have to store the index information to disk as they are processed. Right now I store the index data using pickle, so it would have to be re-written for each new bit of index data. I have looked into converting to an HDF5 index, but the data is not well suited to that format. I will have to think about this a bit further.

For your use case, I would suggest running a small set of test reads with the same reference genome and number of processes to make sure it would run. The memory footprint once the re-squiggle process starts should not change much (in python 2.7 there may still be a memory leak due to cython/numpy partially discontinuing support). Very large reads will add a bit of memory temporarily, but should not be too much compared to the mappy index in each processes. I hope this helps!

**Comment by JohnUrban** [comment]
Well I suppose the clue as to which step is taking a long time hinges on when you have Tombo spit out these messages:
```
Getting file list.
Using default canonical ***** DNA ***** model.
Estimating global scale parameter
```

If file enumeration is starting when it says ```Getting file list.``` and finished before it says ```Using default canonical ***** DNA ***** model.```, then that step is not taking long. There are 8600 fast5 files -- not very many. It is hanging for ~2 days after it prints ```Estimating global scale parameter```. Then it prints ```Correcting 8644 files with 1 subgroup``` and exceeds memory limits soon thereafter.

**Comment by marcus1487** [comment]
Huh. That certainly is hanging on the estimation step. I have seen these status messages delay in logging from a job scheduler, so that is why I thought the messages would not be synced up. 

In any case, that is an extremely long time to open and process 1000 files. I have not seen this step hang for so long in my experience. Are you accessing the files via a remote (maybe long distance) NFS mount?

I know that managing these files can be cumbersome and every file system has its own settings. If possible, it may help to copy the files to some local scratch space to speed up the file access.

**Comment by JohnUrban** [comment]
- RE: "Are you accessing the files via a remote (maybe long distance) NFS mount?"

This seems possible/plausible, although I am not 100% sure how the computing cluster at Brown is set up.

- new Q:

I am guessing "no", but would the ```--include-event-stdev``` or ```--overwrite``` options affect this?

**Comment by marcus1487** [comment]
Hopefully some sort of local data storage could help the first issue, but I probably can't be of much help there.

For the second point, you are correct that neither `--include-event-stdev` nor `--overwrite` should make much of a difference here. `include-event-stdev` will just take a bit longer to compute the standard deviations and the `--overwrite` option will simply skip reads read if data exists in the tombo storage slot already.

**Comment by marcus1487** [comment]
Without the `--overwrite` option set (the default), if re-squiggle results already exist in the `--corrected-group` slot, the read will not be processed and that data will *not* be overwritten. 

If the `--overwrite` option is set, then all reads will be processed whether or not the `--corrected-group` slot in the FAST5 file already contains data. If data does exist in the slot, it will be erased and overwritten with new re-squiggle results.

**Comment by JohnUrban** [comment]
Alrighty - good. That was my original interpretation and how I've been using it. I thought you were saying the opposite in your previous comment so wanted to make sure.

**Comment by marcus1487** [comment]
Yes. Sorry for the confusing language. I think I meant to say, "without the `--overwrite` option reads would be skipped ...".

**Comment by JohnUrban** [comment]
Even though those messages are coming out and it seems to be hanging on ```Estimating global scaling parameter```, I think it is definitely the mapping step that is taking forever.

I just ran it with the default minimap2 index instead of providing my own (w/ more sensitive parameters), and it just finished insanely fast (in comparison to days):
```
Fri Mar  2 14:14:27 EST 2018
resquiggle
Getting file list.
Using default canonical ***** DNA ***** model.
Estimating global scale parameter
Correcting 1799 files with 1 subgroup(s)/read(s) each (Will print a dot for each 1000 reads completed).
.
Failed reads summary (221 total failed):
        Alignment not produced :        152
        Analysis group not found at root of FAST5 :     1
        Not enough raw signal around potential genomic deletion(s) :    7
        Poor raw to expected signal matching :  2
        Poor raw to expected signal matching at read start :    11
        Read event to sequence alignment extends beyond --bandwidth :   48
Fri Mar  2 14:24:57 EST 2018
```

I am guessing the reads are all mapped before it moves on to correction, and that there is no message that states it is in the mapping step after the ```Estimating global scaling parameter``` step.

Any chance I could convince you to add a message such as, ```Mapping reads.``` or ```Anchoring reads to reference sequence(s).```?

**Comment by marcus1487** [comment]
The reads are not mapped before the estimate scale parameter step. Reads are mapped one at a time right before the correction step (following the `Correcting #### files` line). So I am not sure how your command was able to complete so much faster with the new index. My first guess would be that the read information was cached on what ever machine you were using and thus the speed up was actually from the processes not having to access the disk.

**Comment by JohnUrban** [comment]
Perhaps so.... but it is reproducible. The one index always takes days, the other always takes minutes. It is always hanging after the global parameter message. This is the case given the same resources for both. Could it just be that mapping/correction is going on before the correction message? We can chalk it up to a mystery for now if you'd like. This issue is more about the memory demand anyway.

**Comment by JohnUrban** [comment]
Actually - this seems very much to have something to do with providing the index using ```--minimap2-index```. I'm testing that now. I generated the default minimap2 index w/:
```
minimap2 -x map-ont ../genome.fasta -d default-mapont-minimap2.idx
```

It also seems to be hanging at the ``` Estimating global scale parameter``` step -- has been for 6 minutes now. When not providing a minimap2 index, it finishes entirely in <10 minutes. I will keep you updated. If this is the case, I will open a separate issue.

**Comment by JohnUrban** [comment]
It has now started the read correction process -- it moved onto the ```Correcting..``` message after 8-9 minutes or so, then finished 3-4 minutes after that.

It seems to take a little bit longer when providing the default index than just using the default tombo minimap2-indexing -- but nothing like providing the more sensitive minimap2 index.

...so having said that, the only difference is the more sensitive minimap2 index.

**Comment by marcus1487** [comment]
That is good to know that it is not a problem with actually passing the index file, as I was not sure how that could cause this issue. As for using the alternative mapping parameters, have you tried mapping a set of reads with command line minimap2 with this index? Do you see a similar slow down in performance? If the performance is really within the minimap2 mapping code, I'm not sure there is much I can do, and there isn't really a place to add a message for read mapping as it is happening asynchronously across all reads, so it is sort of mixed in with the other correction part of the code.

**Comment by JohnUrban** [comment]
Alright -- I just extracted the fasta reads from the fast5 files, and used them w/ minimap2 directly given the default index or the more sensitive one. With the default mapping finished in 1 second. The more sensitive one has been going for >5 minutes w/o mapping a single read.

So while the more sensitive index (discussed in #28 ) was permissible on a tiny toy reference, it is prohibitive on even small genomes. I will try to find some happy medium.

**Comment by JohnUrban** [comment]
Given that our recent comments came in around the same time, it looks like we were on the same page. Yes -- mapping is the issue. I need to find better parameters.

**Comment by marcus1487** [comment]
Ah good to hear we were on the same page here! And glad we have tracked this down. I'll leave the issue open until I have a fix for the memory issue.

**Comment by marcus1487** [comment]
This issue was fixed in release 1.2.1. This fix involved a bit of a re-write in the mutli-threading scheme for Tombo. The read file IO and mapping (via minimap2/mappy) is completed within separate threads (allowing the minimap2 index to sit in a shared memory space; solving the high memory demand). Then each thread is connected to a single process which completes the computationally intensive re-squiggle part of the algorithm. This new scheme should allow much lower memory consumption, while maintaining the full multi-processing functionality.

This new framework still requires a separate connection to the reference FASTA in order to extract the genomic sequence for an alignment. If the optional dependency pyfaidx is installed this should not result in high memory consumption. Note though that if pyfaidx is not installed one may still encounter large memory demands from the re-squiggle command. An [issues has been raised](https://github.com/lh3/minimap2/issues/126) in mappy to extract the genomic sequence from the minimap2 index. When this functionality is included in a new release it will be included into tombo and unnecessarily large memory consumption should no longer be an issue.
